chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -8,47 +8,160 @@
-BraveDOMWindowStorage::~BraveDOMWindowStorage() {}
+  String session_storage_id = session_storage_namespace->namespace_id() +
+                              String("ephemeral-session-storage");
+  String local_storage_id = session_storage_namespace->namespace_id() +
+                            String("ephemeral-local-storage");
+  supplement = MakeGarbageCollected<EphemeralStorageNamespaces>(
+      StorageController::GetInstance(), session_storage_id, local_storage_id);
+
+  ProvideTo(*page, supplement);
+  return supplement;
+}
 
 // static
-StorageArea* BraveDOMWindowStorage::sessionStorage(LocalDOMWindow& window,
-                                              ExceptionState& exception_state) {
-  auto* storage =
-      DOMWindowStorage::From(window).sessionStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+const char BraveDOMWindowStorage::kSupplementName[] = "DOMWindowStorage";
+
+// static
+BraveDOMWindowStorage& BraveDOMWindowStorage::From(LocalDOMWindow& window) {
+  BraveDOMWindowStorage* supplement =
+      Supplement<LocalDOMWindow>::From<BraveDOMWindowStorage>(window);
+  if (!supplement) {
+    supplement = MakeGarbageCollected<BraveDOMWindowStorage>(window);
+    ProvideTo(window, supplement);
+  }
+  return *supplement;
 }
 
 // static
-StorageArea* BraveDOMWindowStorage::localStorage(LocalDOMWindow& window,
-                                            ExceptionState& exception_state) {
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).sessionStorage(exception_state);
+}
+
+// static
+StorageArea* BraveDOMWindowStorage::localStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).localStorage(exception_state);
+}
+
+BraveDOMWindowStorage::BraveDOMWindowStorage(LocalDOMWindow& window)
+    : Supplement<LocalDOMWindow>(window) {}
+
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    ExceptionState& exception_state) const {
+  if (ephemeral_session_storage_)
+    return ephemeral_session_storage_;
+
+  LocalDOMWindow* window = GetSupplementable();
   auto* storage =
-      DOMWindowStorage::From(window).localStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+      DOMWindowStorage::From(*window).sessionStorage(exception_state);
+
+  if (!window->IsCrossSiteSubframe() || !storage)
+    return storage;
+
+  Page* page = window->GetFrame()->GetDocument()->GetPage();
+  EphemeralStorageNamespaces* namespaces =
+      EphemeralStorageNamespaces::From(page);
+  if (!namespaces)
+    return nullptr;
+
+  Document* document = window->GetFrame()->GetDocument();
+  auto storage_area =
+      namespaces->local_storage()->GetCachedArea(document->GetSecurityOrigin());
+  ephemeral_session_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
```

Comment:
I'm confused about how this works. When does the namespace get added to `storage::SessionStorageImpl`? It seems like it should hit this
```  auto found = namespaces_.find(namespace_id);
  if (found == namespaces_.end()) {
    std::move(callback).Run(/*success=*/false);
    return;
  }```

Reply:
Yes, it seems we need to create `StorageNamespace`s on the browser side. Otherwise, the binds would fail and the data is stored in StoreageAreaMap on the renderer side. That explains why `ClearEphemeralStorage` not effective.
We created a PL to fix this https://github.com/brave/brave-core/pull/6645
PTAL, thanks:)